### PR TITLE
Improve Unit Tests

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**: Improving the unit-tests for deployment, machine-set, and safety-controller.

Fixed a minor bug here:  https://github.com/gardener/machine-controller-manager/pull/567/files#diff-6cb644a14a85dcb91ebd23983896cb2690bfa47af29d0f0daee57f1461cce572R305 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Improved unit-tests for deployment, machineset and safety-controller.
```

### CODE REVIEW

**Review by prashanth26 — COMMENTED** [review]
Although, the main file for `deployment.go` coverage has improved. It would be good if we can think of negative cases to cover the rest of the helper functions in the following files - https://concourse.ci.gardener.cloud/builds/23939#L5fbe03b7:371:476. 

We can do it from this main reconcile method itself with the right set of positive/negative cases - https://github.com/gardener/machine-controller-manager/pull/567/files#diff-83ebd793241bad85de9ff736bc1962bed29e1cb91ecbc610f112c3f2adf14342R1490?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Thanks @hardikdr for the tests. I have given some comments below.

Also, it wasn't clear to me if the cluster-autoscaler integration edge cases for machineset tests were covered. I.e. if the machines annotated by CA are picked first for deletion while scale down. Could you please confirm if this is covered already?

**Comment by hardikdr** [comment]
Thanks @prashanth26 and @amshuman-kr  for the review, I'll address them soon.

File: pkg/controller/machine_safety_test.go [hunk 1]
```diff
@@ -780,4 +780,74 @@ var _ = Describe("#machine_safety", func() {
 		Entry("Control APIServer: UnReachable, Target APIServer: UnReachable, Inactive Timer: Elapsed, Pre-Frozen: true = Post-Frozen: true",
 			false, false, fiveMinutesDuration, true, true),
 	)
+
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/machine_safety_test.go:783]: Sure, I transported the tests to also in OOT safety-net path.

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -711,8 +707,7 @@ func GetFakeMachineFromTemplate(template *v1alpha1.MachineTemplateSpec, parentOb
 		return nil, fmt.Errorf("parentObject does not have ObjectMeta, %v", err)
 	}
 	prefix := getMachinesPrefix(accessor.GetName())
-	rand.Seed(time.Now().UnixNano())
-	prefix = prefix + strconv.Itoa(rand.Intn(100000))
+	prefix = prefix + uuid.New().String()
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/controller_utils.go]: Sure, I capped it with the first 5 letters.

File: pkg/controller/deployment_test.go [hunk 3]
```diff
@@ -0,0 +1,1839 @@
...
+					}
+				},
+				func(testMachineDeployment *machinev1.MachineDeployment, testMachineSets []machinev1.MachineSet) error {
+					if len(testMachineSets) < 1 {
+						return errors.New("It should have created one machine set")
+					}
+					return nil
+				},
+			),
+			Entry("should not create new machineset with dummy-unknown strategy",
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/deployment_test.go]: Sure, I added more cases of recreate, rolling-update, paused, and roll-back.

File: pkg/controller/machine_safety_test.go [hunk 4]
```diff
@@ -780,4 +780,74 @@ var _ = Describe("#machine_safety", func() {
...
+				c.deleteMachineToSafety(testMachine)
+
+				waitForCacheSync(stop, c)
+				Expect(c.machineSafetyOrphanVMsQueue.Len()).To(Equal(expectedQueueSize))
+			},
+			Entry("enqueue the machine key",
+				1,
+			),
+		)
+	})
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/controller/machine_safety_test.go:852]: This method is dependent on functions like `checkAWSMachineClass`, which are completely provider-specific, hence I have left it out in this round.

**Comment by hardikdr** [comment]
>Also, it wasn't clear to me if the cluster-autoscaler integration edge cases for machineset tests were covered. I.e. if the machines annotated by CA are picked first for deletion while scale down. Could you please confirm if this is covered already?

I added the case where the scale-down-disabled-annotation is added on the nodes during rolling-update.

**Comment by hardikdr** [comment]
Integration tests seem to be failing due to unrelated reasons. Unit test execution locally seems to be working fine. 

```••••••••••••••••••••••
Ran 276 of 276 Specs in 120.386 seconds
SUCCESS! -- 276 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
coverage: 48.2% of statements
Apply suggestions from code review

Ginkgo ran 1 suite in 2m11.122842347s
Test Suite Passed
```

**Comment by hardikdr** [comment]
@prashanth26 @amshuman-kr  can you please take a look at the changes if possible?

**Comment by prashanth26** [comment]
Hi @hardikdr,

Missed checking on this. Will check it ASAP.

**Comment by ashwani2k** [comment]
@AxiomSamarth  -- Can we verify this one. Given it being pretty old.
